### PR TITLE
[rush] Fix edge cases where Rush does not update the lockfile

### PR DESCRIPTION
**Pull request by L-Qun**
## Details

image

If the user moves a package from `dependencies` to `devDependencies`, and then runs `rush update`, Rush should update the lockfile correctly, but it does not. This issue occurs because it triggers `case DependencyType.Dev`, but Rush cannot find this package in `importer.devDependencies`. Consequently, Rush deletes it from the `importerDependencies` set.

## Impacted documentation

None.

### CODE REVIEW

File: libraries/rush-lib/src/logic/pnpm/PnpmShrinkwrapFile.ts [hunk 1]
```diff
@@ -1002,6 +1002,7 @@ export class PnpmShrinkwrapFile extends BaseShrinkwrapFile {
             }
             // If fall through, there is a chance the package declares an inconsistent version, ignore it.
             isDevDepFallThrough = true;
+            break;
           }
 
           // eslint-disable-next-line no-fallthrough
```
Comment (iclanton) [inline libraries/rush-lib/src/logic/pnpm/PnpmShrinkwrapFile.ts:1008]:
```suggestion
```

File: libraries/rush-lib/src/logic/pnpm/PnpmShrinkwrapFile.ts [hunk 2]
```diff
@@ -1002,6 +1002,7 @@ export class PnpmShrinkwrapFile extends BaseShrinkwrapFile {
             }
             // If fall through, there is a chance the package declares an inconsistent version, ignore it.
```
Comment (iclanton) [inline libraries/rush-lib/src/logic/pnpm/PnpmShrinkwrapFile.ts:1003]: This should probably be an error case.

File: libraries/rush-lib/src/logic/pnpm/PnpmShrinkwrapFile.ts (hunk 1 above)
Comment (iclanton) [reply libraries/rush-lib/src/logic/pnpm/PnpmShrinkwrapFile.ts:1008]: Given that this explicit rule suppression is here, it seems like there was a reason for this behavior initially. Are you sure unit tests adequately cover all scenarios here?

File: libraries/rush-lib/src/logic/pnpm/PnpmShrinkwrapFile.ts (hunk 2 above)
Comment (iclanton) [reply libraries/rush-lib/src/logic/pnpm/PnpmShrinkwrapFile.ts:1003]: Making this an error case would likely mitigate whatever issue the existing code was trying to account for.

File: common/changes/@microsoft/rush/fix-edge-case-rush-does-not-update-lockfile_2024-08-30-06-44.json [hunk 3]
```diff
@@ -0,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "@microsoft/rush",
+      "comment": "Fix edge cases where Rush does not update the lockfile",
```
Comment (iclanton) [inline common/changes/@microsoft/rush/fix-edge-case-rush-does-not-update-lockfile_2024-08-30-06-44.json:5]:
```suggestion
      "comment": "Fix an issue where moving dependencies from `dependencies` to `devDependencies` in a project's `package.json` did not trigger an lockfile update.",
```

File: libraries/rush-lib/src/logic/pnpm/PnpmShrinkwrapFile.ts (hunk 1 above)
Comment (L-Qun) [reply libraries/rush-lib/src/logic/pnpm/PnpmShrinkwrapFile.ts:1008]: Or maybe I guess it was just an oversight.

**Review by octogonz-dev — APPROVED** [review]

File: libraries/rush-lib/src/logic/pnpm/PnpmShrinkwrapFile.ts (hunk 1 above)
Comment (octogonz-dev) [reply libraries/rush-lib/src/logic/pnpm/PnpmShrinkwrapFile.ts]: @chengcyber and I reviewed this today.  It is really a spec change, not a bug:

In the past, whether a package was a devDependency or regular dependency did not affect the `node_modules` folder. The distinction maybe was reflected in the lockfile, but the installation plan was equivalent.  

But now, apparently `rush install` fails if PNPM finds that the lockfile is inconsistent with **package.json** with regards to devDependencies-vs-dependencies.  We need to modify the spec, such that Rush's analysis now will treat `devDependencies` vs ` dependencies` as a significant difference.

This PR technically fixes the behavior by inserting a `break` statement, however the change is a bit confusing, because the code comments and structure are unchanged.  The /intent/ of the code appears to be the same, but in fact we have changed it.

Possible improvements:
1. Add some code comments better explaining the expected behavior
2. Possibly restructure the code to no longer rely on case statements that fallthrough
3. Ideally add some unit tests with examples of the edge cases

File: libraries/rush-lib/src/logic/pnpm/PnpmShrinkwrapFile.ts (hunk 2 above)
Comment (octogonz-dev) [reply libraries/rush-lib/src/logic/pnpm/PnpmShrinkwrapFile.ts:1003]: > Making this an error case would likely mitigate whatever issue the existing code was trying to account for.

I'm not sure about that. If the inconsistency is in an indirect dependency, and PNPM tolerates it, then Rush probably should as well.
